src/core/jsproxy.c

```diff
@@ -128,45 +130,67 @@ JsProxy_SetAttr(PyObject* self, PyObject* attr, PyObject* pyvalue)
+
 static PyObject*
-JsProxy_Call(PyObject* self, PyObject* args, PyObject* kwargs)
-{
-  bool success = false;
-  JsRef idargs = NULL;
-  JsRef idarg = NULL;
-  JsRef idkwargs = NULL;
-  JsRef idresult = NULL;
-  // result:
-  PyObject* pyresult;
-  Py_ssize_t nargs = PyTuple_Size(args);
+JsProxy_Vectorcall(PyObject* self,
+                   PyObject* const* args,
+                   size_t nargsf,
+                   PyObject* kwnames)
+{
+  bool kwargs = false;
+  if (kwnames != NULL) {
+    // There were kwargs? But maybe kwnames is the empty tuple?
+    PyObject* kwname = PyTuple_GetItem(kwnames, 0);
+    // Clear IndexError
+    PyErr_Clear();
+    if (kwname != NULL) {
+      kwargs = true;
+      if (JsProxy_SUPPORTS_KWARGS(self) == -1) {
+        JsProxy_SUPPORTS_KWARGS(self) =
+          hiwire_function_supports_kwargs(JsProxy_JSREF(self));
+      }
+    }
+    if (kwargs && !JsProxy_SUPPORTS_KWARGS(self)) {
+      // We have kwargs but function doesn't support them. Raise error.
+      const char* kwname_utf8 = PyUnicode_AsUTF8(kwname);
+      PyErr_Format(PyExc_TypeError,
+                   "jsproxy got an unexpected keyword argument '%s'",
+                   kwname_utf8);
+      return NULL;
+    }
+  }
 
-  idargs = hiwire_array();
+  Py_EnterRecursiveCall(" in JsProxy_Vectorcall");
+
+  Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+  JsRef idargs = hiwire_array();
   for (Py_ssize_t i = 0; i < nargs; ++i) {
-    idarg = python2js(PyTuple_GET_ITEM(args, i));
-    FAIL_IF_NULL(idarg);
-    FAIL_IF_MINUS_ONE(hiwire_push_array(idargs, idarg));
-    hiwire_CLEAR(idarg);
+    JsRef idarg = python2js(args[i]);
+    hiwire_push_array(idargs, idarg);
+    hiwire_decref(idarg);
   }
 
-  if (PyDict_Size(kwargs)) {
-    idkwargs = python2js(kwargs);
-    FAIL_IF_MINUS_ONE(hiwire_push_array(idargs, idkwargs));
+  if (kwargs) {
+    // store kwargs into an object which we'll use as the last argument.
+    JsRef idkwargs = hiwire_object();
+    Py_ssize_t nkwargs = PyTuple_Size(kwnames);
+    for (Py_ssize_t i = 0, k = nargsf; i < nkwargs; ++i, ++k) {
+      PyObject* name = PyTuple_GET_ITEM(kwnames, i);
+      const char* name_utf8 = PyUnicode_AsUTF8(name);
+      JsRef jsarg = python2js(args[k]);
+      hiwire_set_member_string(idkwargs, name_utf8, jsarg);
+    }
+    hiwire_push_array(idargs, idkwargs);
+    hiwire_decref(idkwargs);
   }
 
-  idresult = hiwire_call(JsProxy_REF(self), idargs);
-  FAIL_IF_NULL(idresult);
-  pyresult = js2python(idresult);
-  FAIL_IF_NULL(pyresult);
-
-  success = true;
-finally:
-  hiwire_CLEAR(idargs);
-  hiwire_CLEAR(idarg);
-  hiwire_CLEAR(idkwargs);
-  hiwire_CLEAR(idresult);
-  if (!success) {
-    Py_CLEAR(pyresult);
-  }
+  JsRef idresult = hiwire_call(JsProxy_JSREF(self), idargs);
+  hiwire_decref(idargs);
+  PyObject* pyresult = js2python(idresult);
+  hiwire_decref(idresult);
+  Py_LeaveRecursiveCall(/* " in JsProxy_Vectorcall" */);
```

Comment:
We are no longer doing any error checking in this function?

Reply:
This is more or less a merge / sequencing mistake on my part. I wrote this while `error_handling.h` was under work and I meant to throw in a comment saying to come back and add the error handling when that branch was merged, I apparently forgot. Thanks for pointing that out, I will fix it!